### PR TITLE
[TKW] Teach expansion to handle non direct acc and ReduceOp on reduction dim.

### PR DESCRIPTION
**Pull request by raikonenfnu**
In flash attention, we need to enable non direct acc matmul, and also expansion of reduceOp in reduction dimension. The former is needed in FA since we are applying some scaling to the acc of second MMA before feeding it in. The second case is required in FA because ReduceOp/MaxOp is in the backward slice of second MMA's LHS, which would require it to be expanded in K2/reduction dim as well.

### CODE REVIEW

File: iree/turbine/kernel/ops/wave_ops.py [hunk 1]
```diff
@@ -409,11 +409,15 @@ def copy(
             new_node.reduction_dim = self.fx_node.reduction_dim
         return get_custom(new_node)
 
-    def replace_all_uses_with(self, new_node: CustomOp | fx.Node):
+    def replace_all_uses_with(
+        self,
+        new_node: CustomOp | fx.Node,
+        filter_fn: Callable[[fx.node], bool] = lambda x: True,
+    ):
```
Comment (harsh-nod) [inline iree/turbine/kernel/ops/wave_ops.py:416]: Can you add a comment describing what the filter function means? Does it mean that we will replace all uses except the ones that the filter function returns false for?

File: iree/turbine/kernel/ops/wave_ops.py [hunk 2]
```diff
@@ -611,7 +615,7 @@ def type(self) -> Memory:
             raise ValueError(
                 "BinaryPyOp requires lhs and rhs shape to be at least broadcastable."
             )
-        broadcasted_type = lhs_type if lhs_dim_set > rhs_dim_set else rhstype
+        broadcasted_type = lhs_type if lhs_dim_set > rhs_dim_set else rhs_type
```
Comment (harsh-nod) [inline iree/turbine/kernel/ops/wave_ops.py:618]: Nice! Not sure how that got by code review :)

File: iree/turbine/kernel/wave/expansion.py [hunk 3]
```diff
@@ -553,6 +553,53 @@ def get_dim_scaling(
...
+            trace,
+            dim_query,
+            dim_scaling,
+            context,
+            get_node_dim_scaling,
+            res_idx,
+        )
+
+        # This expansion always happens, user should never be reused
+        assert new_node != latest_reduced_op
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/expansion.py:593]: I would modify this comment to say that the node is always cloned and that is why new node will never be equal to latest reduced op

File: iree/turbine/kernel/wave/expansion.py [hunk 4]
```diff
@@ -553,6 +553,53 @@ def get_dim_scaling(
...
+    mma: MMA,
+    trace: CapturedTrace,
+    dim_query: dict[IndexSymbol, int],
+    dim_scaling: dict[IndexSymbol, int],
+    context: ExpandedNodeMap,
+    get_node_dim_scaling: Callable[[fx.Node], dict[IndexSymbol, int]],
+    res_idx: int,
+) -> CustomOp:
+    latest_reduced_op = mma
+    for scale_idx in range(1, dim_scaling[mma.reduction_dim]):
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/expansion.py:566]: Just wondering why this has to start from 1?

File: iree/turbine/kernel/wave/expansion.py [hunk 5]
```diff
@@ -566,53 +613,63 @@ def _handle_reduction_dim(
...
-
-                saved_arg = user.node_args[index]
-                user.update_arg(index, dummy)
-                new_node = _expand_node(
-                    user,
+        dim_scaling = get_node_dim_scaling(root_op)
+        dims = dict(root_op.fx_node.expanded_dims)
+        latest_reduced_op = root_op
+        op_output_index = get_output_index(root_op)
+        if dim_scaling[reduction.axis] <= 1:
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/expansion.py:644]: Do we need this check? Shouldn't it be automatically handled?

File: iree/turbine/kernel/wave/expansion.py [hunk 6]
```diff
@@ -566,53 +613,63 @@ def _handle_reduction_dim(
...
+                dims,
+                dim_scaling,
+                context,
+                get_node_dim_scaling,
+                res_idx,
+            )
+        elif isinstance(root_op, ReduceOp):
+            original_src = latest_reduced_op.arg
+            for scale_idx in range(1, dim_scaling[reduction.axis]):
+                dims[root_op.reduction_dim] = scale_idx
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/expansion.py:659]: this is pretty cool, so if i understand this right, you are expanding all the inputs and replacing the original argument with all the expanded args?

File: iree/turbine/kernel/wave/expansion.py [hunk 7]
```diff
@@ -566,53 +613,63 @@ def _handle_reduction_dim(
     # TODO: Register iter args with the reduction initially so accessing them is easier
     iter_args: list[CustomOp] = []
     reduction_subgraph = trace.get_subgraph(reduction.subgraph_name)
+
+    # TODO: Add support for case where we process MMA before returning to IterArg.
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/expansion.py:617]: Don't fully follow this comment, can you explain?

**Review by harsh-nod — CHANGES_REQUESTED** [review]
Overall, looks good! Thanks I like the changes to expansion. Just some comments about documentation .

File: iree/turbine/kernel/ops/wave_ops.py (hunk 1 above)
Comment (raikonenfnu) [reply iree/turbine/kernel/ops/wave_ops.py:416]: Ah, this is dead code, Used this previously to make `replace_all_uses_with` work more similar to `replace_all_uses_except` where the `filter_fn` is the lambda that skips over op if it's true.

File: iree/turbine/kernel/ops/wave_ops.py (hunk 2 above)
Comment (raikonenfnu) [reply iree/turbine/kernel/ops/wave_ops.py:614]: haha yee :)

File: iree/turbine/kernel/wave/expansion.py (hunk 4 above)
Comment (raikonenfnu) [reply iree/turbine/kernel/wave/expansion.py:569]: Added the doc here, it was from the original code
`The initial nodes are expanded in the first dimension, so we start from 1`

File: iree/turbine/kernel/wave/expansion.py (hunk 7 above)
Comment (raikonenfnu) [reply iree/turbine/kernel/wave/expansion.py]: Cleared up the comment to "Handle case where MMAs/ReduceOps do not have Output as direct consumer."
LMK if it's still not well worded :)

File: iree/turbine/kernel/wave/expansion.py (hunk 5 above)
Comment (raikonenfnu) [reply iree/turbine/kernel/wave/expansion.py]: removed

File: iree/turbine/kernel/wave/expansion.py (hunk 6 above)
Comment (raikonenfnu) [reply iree/turbine/kernel/wave/expansion.py:660]: That is correct!

File: iree/turbine/kernel/wave/expansion.py (hunk 3 above)
Comment (raikonenfnu) [reply iree/turbine/kernel/wave/expansion.py:596]: done!

File: iree/turbine/kernel/ops/wave_ops.py (hunk 1 above)
Comment (harsh-nod) [reply iree/turbine/kernel/ops/wave_ops.py]: Yes that's what it looked like to me when I first saw it. Sounds like a useful function to have but we can add that later.

File: iree/turbine/kernel/wave/expansion.py (hunk 5 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/expansion.py]: cool thanks

File: iree/turbine/kernel/wave/expansion.py (hunk 6 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/expansion.py:660]: thanks

File: iree/turbine/kernel/wave/expansion.py (hunk 7 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/expansion.py]: yes that makes a lot more sense, thanks

File: iree/turbine/kernel/wave/expansion.py [hunk 8]
```diff
@@ -327,33 +327,35 @@ def _expand_reduction(
...
+                arg,
+                trace,
+                dims,
+                get_node_dim_scaling(arg),
+                context,
+                get_node_dim_scaling,
+                res_idx,
             )
+            # Proceed with expansion inside the reduction
+            if expanded_output in new_output_args:
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/expansion.py:340]: when do you get this situation where you expanded something that is already in the new_output_args? is it even worth expanding if it is not needed? can we use the context to detect this and prevent expanding?

File: iree/turbine/kernel/ops/wave_ops.py [hunk 9]
```diff
@@ -1300,7 +1300,12 @@ def type(self) -> Memory:
             from ..wave.utils import all_equal
 
             src_types = [get_custom(arg).type for arg in self.arg]
-            if not all_equal(src_types):
+            ref_shape = src_types[0].symbolic_shape
```
[the remark this answers is not in the document]
Comment (raikonenfnu) [reply iree/turbine/kernel/ops/wave_ops.py:1303]: For some reason, I have cases where `Register[M, K2].of(f32)) != `Register[M, K2].of(f32))`. I think something to do with the metaclass

File: iree/turbine/kernel/wave/expansion.py (hunk 8 above)
Comment (raikonenfnu) [reply iree/turbine/kernel/wave/expansion.py:340]: It's not getting re-expanded, but for some reason it's getting reinserted

Comment (harsh-nod) [reply iree/turbine/kernel/wave/expansion.py:340]: same as above, maybe add a TODO explaining this and we can look into this later.

**Review by harsh-nod — APPROVED** [review]
overall looks good and we can address the open questions later. Thanks!

File: iree/turbine/kernel/ops/wave_ops.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (raikonenfnu) [reply iree/turbine/kernel/ops/wave_ops.py:1303]: done! :)

File: iree/turbine/kernel/wave/expansion.py (hunk 8 above)
Comment (raikonenfnu) [reply iree/turbine/kernel/wave/expansion.py:352]: done!

File: iree/turbine/kernel/ops/wave_ops.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (harsh-nod) [reply iree/turbine/kernel/ops/wave_ops.py:1303]: thanks!

**Pull request merged**